SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -81,72 +142,240 @@ abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
+  have H : ∃ (k : (ZMod q)[X]), (R.fromPoly (n:=n) a).representative' = k * (f q n) + a := by apply
+    R.fromPoly_rep'_eq_element;
+  obtain ⟨k, hk⟩ := H
+  exists k
+  ring_nf at hk ⊢
+  rw[← hk]
+  norm_cast
+
+/-- The representative of 0 wil live in the ideal of {f q n}. To show that such an element is a multiple of {f q n}, use `Ideal.mem_span_singleton'`-/
+theorem R.representative'_zero_ideal : R.representative' q n 0 ∈ Ideal.span {f q n} := by
+  have H : ∃ i ∈ Ideal.span {f q n}, (R.fromPoly (n:=n) 0).representative' = 0 + i := by apply R.fromPoly_rep'_eq_ideal (a := 0)
+  obtain ⟨i, hi, hi'⟩ := H
+  simp[fromPoly] at hi'
+  rw[hi']
+  assumption
+
+/-- The representatiatve of 0 is a multiple of `f q n`. -/
+theorem R.representative'_zero_elem : ∃ (k : (ZMod q)[X]), R.representative' q n 0 = k * (f q n) := by
+  have H : ∃ k : (ZMod q)[X], (R.fromPoly (n:=n) 0).representative' = k * (f q n) + 0 := by apply R.fromPoly_rep'_eq_element (a := 0)
+  obtain ⟨k, hk⟩ := H
+  simp[fromPoly] at hk
+  rw[hk]
+  exists k
+
+/-- Representative of (0 : R) is (0 : Z/qZ[X]) -/
+theorem R.representative_zero : R.representative q n 0 = 0 := by
+  simp[R.representative]
+  obtain ⟨k, hk⟩ := R.representative'_zero_elem q n
+  rw[hk]
+  rw[dvd_iff_modByMonic_eq_zero]
+  simp
+  exact (f_monic q n)
 
 /--
 Characterization theorem for the representative.
 Taking the representative of the equivalence class of a polynomial  `a : (ZMod q)[X]` is
 the same as taking the remainder of `a` modulo `f q n`.
 -/
+theorem R.representative_fromPoly_toFun : forall a : (ZMod q)[X], ((R.fromPoly (n:=n) (q := q)).toFun a).representative = a %ₘ (f q n) := by
+  intro a
+  simp [R.representative]
+  have ⟨i,⟨hiI,hi_eq⟩⟩ := R.fromPoly_rep'_eq_ideal q n a
+  simp[FunLike.coe] at hi_eq
+  rw [hi_eq]
+  apply Polynomial.modByMonic_eq_of_dvd_sub (f_monic q n)
+  ring_nf
+  apply Ideal.mem_span_singleton.1 hiI
+  done
+
 theorem R.representative_fromPoly : forall a : (ZMod q)[X], (R.fromPoly (n:=n) a).representative = a %ₘ (f q n) := by
   intro a
   simp [R.representative]
-  have ⟨i,⟨hiI,hi_eq⟩⟩ := R.fromPoly_rep'_eq q n a
+  have ⟨i,⟨hiI,hi_eq⟩⟩ := R.fromPoly_rep'_eq_ideal q n a
   rw [hi_eq]
   apply Polynomial.modByMonic_eq_of_dvd_sub (f_monic q n)
   ring_nf
   apply Ideal.mem_span_singleton.1 hiI
   done
 
+/-- Representative is an additive homomorphism -/
+@[simp]
+theorem R.representative_add [Fact (q > 1)](a b : R q n) : (a + b).representative = a.representative + b.representative := by
+  have ⟨a', ha'⟩ := R.surjective_fromPoly q n a
+  have ⟨b', hb'⟩ := R.surjective_fromPoly q n b
+  have ⟨ab', hab'⟩ := R.surjective_fromPoly q n (a + b)
+  rw[← hab']
+  subst ha'
+  subst hb'
+  rw[← map_add] at hab'
+  rw[hab']
+  repeat rw[R.representative_fromPoly]
+  rw[Polynomial.add_modByMonic]
+
+/-- pushing and pulling negation through mod -/
+theorem neg_modByMonic (p mod : (ZMod q)[X]) : (-p) %ₘ mod = - (p %ₘ mod) := by
```

Comment:
Are you sure this isn't already there? If it isn't then PR it.

Reply:
https://github.com/leanprover-community/mathlib4/pull/7971